src/power_grid_model/_utils.py

```diff
@@ -299,40 +301,148 @@ def convert_single_dataset_to_python_single_dataset(data: SingleDataset) -> Sing
+        return {}
+    if isinstance(attrs, EllipsisType):
+        names = data.dtype.names if isinstance(data, np.ndarray) else data.keys()
+        return {attr: deepcopy(data[attr]) for attr in names}
+    return {attr: deepcopy(data[attr]) for attr in attrs}
+
+
+def process_data_filter(
+    dataset_type: DatasetType,
+    data_filter: ComponentAttributeMapping,
+    available_components: list[ComponentType],
+) -> _ComponentAttributeMappingDict:
+    """Checks valid type for data_filter. Also checks for any invalid component names and attribute names.
+
+    Args:
+        dataset_type (DatasetType): the type of output that the user will see (as per the calculation options)
+        data_filter (ComponentAttributeMapping):  data_filter provided by user
+        available_components (list[ComponentType]):  all components available in model instance or data
+
+    Returns:
+        _ComponentAttributeMappingDict: processed data_filter in a dictionary
+    """
+    if data_filter is None:
+        processed_data_filter: _ComponentAttributeMappingDict = {ComponentType[k]: None for k in available_components}
+    elif data_filter is Ellipsis:
+        processed_data_filter = {ComponentType[k]: ... for k in available_components}
+    elif isinstance(data_filter, (list, set)):
+        processed_data_filter = {ComponentType[k]: None for k in data_filter}
+    elif isinstance(data_filter, dict) and all(
+        attrs is None or attrs is Ellipsis or isinstance(attrs, (set, list)) for attrs in data_filter.values()
+    ):
+        processed_data_filter = data_filter
+    else:
+        raise ValueError(f"Invalid filter provided: {data_filter}")
+
+    validate_data_filter(processed_data_filter, dataset_type, available_components)
+    return processed_data_filter
+
+
+def validate_data_filter(
+    data_filter: _ComponentAttributeMappingDict, dataset_type: DatasetType, available_components: list[ComponentType]
+) -> None:
+    """Raise error if some specified components or attributes are unknown.
+
+    Args:
+        data_filter (_ComponentAttributeMappingDict): Processed component to attribtue dictionary
+        dataset_type (DatasetType):  Type of dataset
+        available_components (list[ComponentType]):  all components available in model instance or data
+
+    Raises:
+        ValueError: when the type for data_filter is incorrect
+        KeyError: with "unknown component types" for any unknown components
+        KeyError: with "unknown attributes" for unknown attribute(s) for a known component
+    """
+    dataset_meta = power_grid_meta_data[dataset_type]
+
+    for source, components in {"data_filter": data_filter.keys(), "data": available_components}.items():
+        unknown_components = [x for x in components if x not in dataset_meta]
+        if unknown_components:
+            raise KeyError(f"You have specified some unknown component types: {unknown_components} in {source}")
```

Comment:
better to be explicit:
```suggestion
            raise KeyError(f"The following specified component types are unknown:{unknown_components} in {source}")
```
idem below

Reply:
The existing error message is f"You have specified some unknown component types: {unknown_components}"
Maybe its better to conform with that